categorize_notify/program.py

```diff
@@ -0,0 +1,162 @@
+"""
+This program demonstrates a real-life workflow that integrates Gmail, ChatGPT, and Slack.
+
+Workflow:
+1. Trigger: Detect a new email in Gmail.
+2. Categorize: Use ChatGPT to read and categorize the email (e.g., technical work, marketing, sales).
+3. Notify: Send a message to the corresponding Slack channel based on the category.
+4. Label: Add a label to the email in Gmail.
+"""
+
+import base64
+import os
+import time
+
+import autokitteh
+from autokitteh import google, slack
+from openai import OpenAI
+
+
+GMAIL_CONNECTION_NAME = "my_gmail"
+OPENAI_CONNECTION_NAME = "my_chatgpt"
+SLACK_CHANNELS = ["demos", "engineering", "ui"]
+SLACK_CONNECTION_NAME = "my_slack"
+
+
+def on_http_get(data):
+    total_messages = None
+    while True:
+        total_messages = _poll_inbox(total_messages)
+        time.sleep(10)
+
+
+@autokitteh.activity
+def _poll_inbox(prev_total_messages: int):
+    gmail = google.gmail_client(GMAIL_CONNECTION_NAME).users()
+    # TODO: fix: If user deletes emails before the next poll then new email messages will be missed.
+    curr_total_messages = _get_message_count(gmail)
+    if prev_total_messages and curr_total_messages > prev_total_messages:
+        new_email_count = curr_total_messages - prev_total_messages
+        message_ids = _get_latest_message_ids(gmail, new_email_count)
+        for message_id in message_ids:
+            _process_email(gmail, message_id)
+
+    return curr_total_messages
+
+
+def _process_email(gmail, message_id: str):
+    message = gmail.messages().get(userId="me", id=message_id).execute()
+    email_content = _parse_email(message)
+    if email_content:
+        channel = _categorize_email(email_content, SLACK_CHANNELS)
+        if channel:
+            client = slack.slack_client(SLACK_CONNECTION_NAME)
+            client.chat_postMessage(channel=channel, text=email_content)
+
+        # Add label to email
+        label_id = _get_label_id(gmail, channel)
+        if not label_id:
+            created_label = _create_label(gmail, channel)
+            label_id = created_label["id"]
```

Comment:
Why not make `_create_label()` return just the label instead of a dict?
If you do that, you can replace these 4 lines with:

```
label_id = _get_label_id(gmail, channel) or _create_label(gmail, channel)
```